packages/server/test/integration/conduits.test.ts

```diff
@@ -0,0 +1,162 @@
+      channelConfig: { botToken: randStr() }
+    }
+  })
+
+  afterAll(async () => {
+    await app.destroy()
+  })
+
+  beforeEach(async () => {
+    app.caching.resetAll()
+  })
+
+  test('Create conduit', async () => {
+    const conduit = await conduits.create(state.provider.id, state.channel.meta.id, state.channelConfig)
+
+    expect(conduit).toBeDefined()
+    expect(validateUuid(conduit.id)).toBeTruthy()
+    expect(conduit.channelId).toEqual(state.channel.meta.id)
+    expect(conduit.providerId).toEqual(state.provider.id)
+    expect(conduit.config).toEqual(state.channelConfig)
+
+    state.conduit = conduit
+  })
+
+  test('Creating same conduit throws', async () => {
+    await expect(conduits.create(state.provider.id, state.channel.meta.id, state.channelConfig)).rejects.toThrow()
+  })
+
+  test('Creating conduit with invalid config throws', async () => {
+    const otherProvider = await app.providers.create(randStr(), false)
+    await expect(conduits.create(otherProvider.id, state.channel.meta.id, { yo: randStr() })).rejects.toThrow()
+  })
+
+  test('Get conduit by id', async () => {
+    const conduit = await conduits.get(state.conduit!.id)
+    expect(conduit).toEqual(state.conduit)
+    expect(querySpy).toHaveBeenCalledTimes(1)
+  })
+
+  test('Get conduit by id cached', async () => {
+    const conduit = await conduits.get(state.conduit!.id)
+    expect(conduit).toEqual(state.conduit)
+    expect(querySpy).toHaveBeenCalledTimes(1)
+
+    for (let i = 0; i < 10; i++) {
+      const cached = await conduits.get(state.conduit!.id)
+      expect(cached).toEqual(state.conduit)
+    }
+
+    expect(querySpy).toHaveBeenCalledTimes(1)
+  })
+
+  test('Get conduit by provider and channel', async () => {
+    const conduit = await conduits.getByProviderAndChannel(state.provider.id, state.channel.meta.id)
+    expect(conduit).toEqual(state.conduit)
+    expect(querySpy).toHaveBeenCalledTimes(1)
+  })
+
+  test('Get conduit by provider and channel cached', async () => {
+    const conduit = await conduits.getByProviderAndChannel(state.provider.id, state.channel.meta.id)
+    expect(conduit).toEqual(state.conduit)
+    expect(querySpy).toHaveBeenCalledTimes(1)
+
+    for (let i = 0; i < 10; i++) {
+      const cached = await conduits.getByProviderAndChannel(state.provider.id, state.channel.meta.id)
+      expect(cached).toEqual(state.conduit)
+    }
+
+    expect(querySpy).toHaveBeenCalledTimes(1)
+  })
+
+  test('Updating conduit config clears cache and persist changes', async () => {
+    const newConfig = { botToken: randStr() }
+    await conduits.updateConfig(state.conduit!.id, newConfig)
+    const calls = querySpy.mock.calls.length
+
+    const conduit = await conduits.get(state.conduit!.id)
+    expect(conduit).toEqual({ ...state.conduit, config: newConfig })
+    // TODO: doesn't work because this is reacted to in an event. Should events be disabled in tests?
+    // expect(querySpy).toHaveBeenCalledTimes(calls + 1)
```

Comment:
You could maybe spy on the event dispatcher too then?

Reply:
We don't really need to check that here anyways so I removed it.